### ISSUE TITLE
Sequence Prediction [info request]

### ISSUE THREAD

**Issue by eblabs** [Unity-Technologies/ml-agents#1573]
Hi,
I'm getting a lot out of ml-agents, really great work to the team! Just have a question about how to approach sequence prediction using ml-agents.

I've created this GRU based RNN in python/keras/tensorflow before attempting to build this in Unity. It is intended to predict the pattern/timing of footsteps based on the character's speed. Its trained on data extracted from a long mocap clip.

It has a sequence of 3 inputs;
previous-frame character speed
previous-frame left foot Y
previous-frame right foot Y

And it predicts a sequence of 2 outputs;
current-frame left foot Y
current-frame right foot Y

![predictiongraph](https://user-images.githubusercontent.com/5941510/50782459-eaf92b00-125c-11e9-90a5-aeae81af6635.png)
* you can see the left and right feet moving up/down as the player's speed increases then return to the ground (0) once the player stops moving.

Since the intent of this NN is to generate a pattern/sequence based on the history of a players speed and previous footsteps, it needs to be trained with this mocap data. I'm not sure how this fits into the control-based/reward paradigm of ml-agent training. Maybe this could work through ml-agent's Imitation Learning, but this seems to be focused on recording input from a player rather than some arbitrary training data. 

At any rate, if there were any broad suggestions on how to approach this from within Unity that would be great.

Thanks everyone!!

**Comment by eblabs** [comment]
So I'm making a bit of progress with the Heuristic Brain. I'm basically using the Decision script to feedback in the 'y' portion of the training data on the Teacher Agent. I found it quite slow doing the training in online_bc, both with and without the --slow flag. I tried the offline_bc method and it looks like a demo recorder is required. Although with a Heuristic Brain, I didn't think the demo recorder would be always be necessary with this x/y training scenario where all the training data is known ahead of time.

Thanks for any thoughts on this!

**Comment by mmattar** [comment]
@eblabs - just to clarify, it seems that you're just using Unity to generate the data and send it off to Python for training and that the training process in TF does not actually influence the Unity scene. Is that correct? If so, why not just record a large data set from a Unity run and save the data to file and then run the training offline. You could use the demonstration recorder, but also something that just prints out a csv file.

**Comment by harperj** [comment]
@eblabs This sounds like a reasonable approach, though I can't speak to how the behavior of our behavioral cloning trainer will compare to the model you used previously.  Hopefully this worked out for you in the end.

It has been a while since any activity on this issue so I'm going to close it, but if you have any more to share or have run into any problems please feel free to reopen.

**Comment by eblabs** [comment]
Thanks @harperj!
Things didnt quite work out how I had hoped. The offline training method ended up being much slower for some reason, and I couldn't quite achieve the same results as with using Tensor/Karas in python. 

I had considered exporting the network+data from python into Unity, but I didn't have the time to get into this as from what I had read, it wasn't totally straightforward. (having correctly formatted, proper input naming, etc)

It seems like some tasks are better suited to the Unity training method. It would be nice to have some kind of guide for generating models and data outside of unity and bringing them in.

Thanks again!
